File: part2/protocols/bin_file_write.cpp
// Apache Thrift Binary Protocol Struct Writer in C++ 

#include <thrift/transport/TSimpleFileTransport.h>
#include <thrift/protocol/TProtocol.h>
#include <thrift/protocol/TBinaryProtocol.h>
#include <boost/make_shared.hpp>
#include <iostream>
#include <string>

using namespace apache::thrift::transport;
using namespace apache::thrift::protocol;
using boost::make_shared;

struct Trade {
   char symbol [16];
   double price;
   int size;
};

int main()
{
    auto path_name = "data";
    auto trans = make_shared<TSimpleFileTransport>(path_name, false, true);
    auto proto = make_shared<TBinaryProtocol>(trans);

    Trade trade{"F", 13.10, 2500};
    int i = 0;
    i += proto->writeStructBegin("Trade");

    i += proto->writeFieldBegin("symbol ",
                               ::apache::thrift::protocol::T_STRING, 1);
    i += proto->writeString(std::string(trade.symbol));
    i += proto->writeFieldEnd();

    i += proto->writeFieldBegin("price ",
                               ::apache::thrift::protocol::T_DOUBLE, 2);
    i += proto->writeDouble(trade.price);
    i += proto->writeFieldEnd();

    i += proto->writeFieldBegin("size ", ::apache::thrift::protocol::T_I32, 3);
    i += proto->writeI32(trade.size);
    i += proto->writeFieldEnd();

    i += proto->writeFieldStop();
    i += proto->writeStructEnd();

    std::cout << "Wrote " << i << " bytes to " << path_name << std::endl;
}

File: part2/protocols/bin_file_write.py
# Apache Thrift Binary Protocol Struct 2.0 Writer in Python

from thrift.transport import TTransport
from thrift.protocol import TBinaryProtocol
from thrift import Thrift

class Trade:
    def __init__(self, symbol="", price=0.0, size=0, timestamp=0.0):
        self.symbol=symbol
        self.price=price
        self.size=size
        self.timestamp=timestamp

trans = TTransport.TFileObjectTransport(open("data","wb"))
proto = TBinaryProtocol.TBinaryProtocol(trans)

trade = Trade("GE", 27.25, 1700, 9.5)

proto.writeStructBegin("Trade")

proto.writeFieldBegin("symbol", Thrift.TType.STRING, 1)
proto.writeString(trade.symbol)
proto.writeFieldEnd()

proto.writeFieldBegin("price", Thrift.TType.DOUBLE, 2)
proto.writeDouble(trade.price)
proto.writeFieldEnd()

proto.writeFieldBegin("size", Thrift.TType.I32, 3)
proto.writeI32(trade.size)
proto.writeFieldEnd()

proto.writeFieldBegin("timestamp", Thrift.TType.DOUBLE, 4)	
proto.writeDouble(trade.timestamp)
proto.writeFieldEnd()

proto.writeFieldStop()
proto.writeStructEnd()
        
print("Wrote Trade: %s %d @ %f  tm: %f" % 
      (trade.symbol, trade.size, trade.price, trade.timestamp))

